<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Откройте-таблицу-и-изучите-общую-информацию-о-данных" data-toc-modified-id="Откройте-таблицу-и-изучите-общую-информацию-о-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Откройте таблицу и изучите общую информацию о данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-пропусков" data-toc-modified-id="Удаление-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Удаление пропусков</a></span></li><li><span><a href="#Обработка-аномальных-значений" data-toc-modified-id="Обработка-аномальных-значений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка аномальных значений</a></span></li><li><span><a href="#Удаление-пропусков-(продолжение)" data-toc-modified-id="Удаление-пропусков-(продолжение)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Удаление пропусков (продолжение)</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация данных</a></span></li><li><span><a href="#Шаг-3.-Исследуйте-данные-и-ответьте-на-вопросы" data-toc-modified-id="Шаг-3.-Исследуйте-данные-и-ответьте-на-вопросы-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Шаг 3. Исследуйте данные и ответьте на вопросы</a></span></li><li><span><a href="#Шаг-3.-Исследуйте-данные-и-ответьте-на-вопросы" data-toc-modified-id="Шаг-3.-Исследуйте-данные-и-ответьте-на-вопросы-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>Шаг 3. Исследуйте данные и ответьте на вопросы</a></span><ul class="toc-item"><li><span><a href="#3.1-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="3.1-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-2.8.1"><span class="toc-item-num">2.8.1&nbsp;&nbsp;</span>3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#3.2-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="3.2-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-2.8.2"><span class="toc-item-num">2.8.2&nbsp;&nbsp;</span>3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#3.3-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="3.3-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-2.8.3"><span class="toc-item-num">2.8.3&nbsp;&nbsp;</span>3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#3.4-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="3.4-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-2.8.4"><span class="toc-item-num">2.8.4&nbsp;&nbsp;</span>3.4 Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#3.5-Приведите-возможные-причины-появления-пропусков-в-исходных-данных." data-toc-modified-id="3.5-Приведите-возможные-причины-появления-пропусков-в-исходных-данных.-2.8.5"><span class="toc-item-num">2.8.5&nbsp;&nbsp;</span>3.5 Приведите возможные причины появления пропусков в исходных данных.</a></span></li><li><span><a href="#3.6-Объясните,-почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных." data-toc-modified-id="3.6-Объясните,-почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных.-2.8.6"><span class="toc-item-num">2.8.6&nbsp;&nbsp;</span>3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.</a></span></li></ul></li><li><span><a href="#Шаг-4:-общий-вывод." data-toc-modified-id="Шаг-4:-общий-вывод.-2.9"><span class="toc-item-num">2.9&nbsp;&nbsp;</span>Шаг 4: общий вывод.</a></span></li></ul></li></ul></div>

# Исследование надежности заемщиков


 Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.
 
 **Цели:** Определить влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.  
 
 **Предоставленные данные:** Входные данные от банка — статистика о платёжеспособности клиентов.

Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

In [2]:
try:
    data = pd.read_csv('/datasets/data.csv') #
except:
    data = pd.read_csv('data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Краткий вывод:** Осуществили выгрузку базы, вывели общую информацию. Таблица состоит из 21525 строк и 12 столбцов. Столбцы "days_employed" и "total_income" имеют пропуски, они будут обработаны в следующем разделе.

## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

**Краткий вывод:** Были обработаны пропуски в столбцах "days_employed" и "total_income", аномальные значения в стлбце "сhildren", обработаны дубликаты в столбце "education". Обработаны столбцы total_income и purpose, и выполнена их категоризация.

### Шаг 3. Исследуйте данные и ответьте на вопросы

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [23]:
children_debt = data.groupby('children')['debt'].agg(['count', 'mean']).sort_values(by = 'mean', ascending = False)

print ('Процент должников по количеству детей:')
print (children_debt)


Процент должников по количеству детей:
          count      mean
children                 
4            41  0.097561
2          2052  0.094542
1          4808  0.092346
3           330  0.081818
0         14091  0.075438
5             9  0.000000


**Вывод:** Из таблицы видно, что среди кредитополучателей без детей, должников намного меньше. С уверенностью можно сказать, что  семьи с 1 и 2 детьми задерживают оплату чаще. Многодетные семьи так же задерживают оплату, но одназначного ответа по этим категориям дать нельзя, по причине малой выборки. 

In [24]:
children_debt

count      mean
children                 
4            41  0.097561
2          2052  0.094542
1          4808  0.092346
3           330  0.081818
0         14091  0.075438
5             9  0.000000

In [25]:
children_debt.style.background_gradient(subset = 'mean', vmin=0.07, vmax=0.1)

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [26]:
fstatus_debt = data.pivot_table(index = 'family_status', values = 'debt', aggfunc = ['count', 'mean'])\
    .sort_values(by = ('mean', 'debt'), ascending = False)
print ('Процент должников в зависимости от семейного статуса:')
print (fstatus_debt)

Процент должников в зависимости от семейного статуса:
                       count      mean
                        debt      debt
family_status                         
Не женат / не замужем   2796  0.097639
гражданский брак        4134  0.093130
женат / замужем        12261  0.075606
в разводе               1189  0.070648
вдовец / вдова           951  0.066246


**Вывод:** Среди тех кто состоит или состояли в браке, процент должников значительно меньше.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [27]:
diap = {'total_income_cat_diap' : ['0–30000', '30001–50000', '50001–200000','200001–1000000', '1000001 и выше'],\
        'total_income_category' : ['E','D','C','B','A']}
df_diap = pd.DataFrame(diap)

income_debt = data.pivot_table(index = 'total_income_category', values = 'debt', aggfunc = ['count', 'mean'])

income_debt_result = df_diap.merge(income_debt, on='total_income_category', how='left')\
    .sort_values(by = ('mean', 'debt'), ascending = False)
print ('Процент должников в зависимости от уровня дохода:')
print (income_debt_result)



Процент должников в зависимости от уровня дохода:
  total_income_cat_diap total_income_category  (count, debt)  (mean, debt)
0               0–30000                     E             22      0.090909
2          50001–200000                     C          15921      0.084982
4        1000001 и выше                     A             25      0.080000
3        200001–1000000                     B           5014      0.070602
1           30001–50000                     D            349      0.060172


C:\Temp\ipykernel_9064\4288466819.py:7: FutureWarning: merging between different levels is deprecated and will be removed in a future version. (1 levels on the left, 2 on the right)
  income_debt_result = df_diap.merge(income_debt, on='total_income_category', how='left')\


**Вывод:** С уверенностью можно сделать вывод лишь по двум категориям, (C и B), с увеличением дохода процент должников по этим категориям падает. Для полноценного анализа завсимости, необходимо иметь большую выборку по всем категориям.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [28]:
purpose_debt = data.groupby('purpose_category')['debt'].agg(['count', 'mean']).sort_values(by = 'mean', ascending = False)

print ('Процент должников в зависимости от цели взятия кредита:')
print (purpose_debt)

Процент должников в зависимости от цели взятия кредита:
                          count      mean
purpose_category                         
операции с автомобилем     4279  0.093480
получение образования      3988  0.092528
проведение свадьбы         2313  0.079118
операции с недвижимостью  10751  0.072551


**Вывод**: Должников среди тех, кто берет кредит на покупку недвижимости или проведение свадьбы, значительно меньше, чем в других случаях.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Причины: 1. Человеческий фактор.(Забыли заполнить поле, заполнили не верно, итд.) 2. Техническая ошибка (ошибка при выгрузке, неправильный алгоритм выгрузки)

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Среднее значение может сильно искажать статистику в большую или меньшую сторону, при наличии очень больших\маленьких значений. Медиана лишена этого недостатка т.к. используется серединное значение всего диапазона.

### Шаг 4: общий вывод.

    В ходе проведенной работы, предоставленная банком информация, база данных должников кредитного отдела была соответствующим образом обработана (обработаны пропуски, дубликаты, данные приведены в надлежащий вид, произведена категоризация).  
        Банком была поставлена задача: определить влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.
    - Семейное положение влияет на погашения кредита в срок. Среди людей не состоявших в браке (неженатые\незамужние\гражданский брак ) в среднем должников на 2% больше, чем люди состоявшие в браке.
    
    - Количество детей так же влияет на погашения кредита в срок, семьи без детей реже становяться должникам, чем семьи с детьми. Среди кредитополучателей без детей, должников намного меньше (примерно на 2%). С уверенностью можно сказать, что семьи с 1 и 2 детьми задерживают оплату чаще. Многодетные семьи(3+) так же задерживают оплату, но одназначного ответа по этим категориям дать нельзя, по причине малой выборки.

